forsys/fmatrix.py

```diff
@@ -93,15 +96,12 @@ def _build_matrix(self) -> Matrix:
 
             if at_least_three and less_than_four_condition:
                 self.map_vid_to_row[vid] = position_index
+                mat[position_index] = row_x
+                mat[position_index + 1] = row_y
                 position_index += 2
-                if len(self.matrix) == 0:
-                    self.matrix = Matrix(( [row_x] ))
-                    self.matrix = self.matrix.row_insert(self.matrix.shape[0], Matrix(( [row_y] )))
-                else:
-                    self.matrix = self.matrix.row_insert(self.matrix.shape[0], Matrix(([row_x])))
-                    self.matrix = self.matrix.row_insert(self.matrix.shape[0], Matrix(([row_y])))
 
-        return self.matrix
+        # return matrix with populated rows
+        return mat[:position_index]

```

Comment:
Why not return the complete matrix? Shouldn't it be the correct size that you created in L87?

Reply:
From what I've understood, 2 x n_vertices is the maximum number of rows that it can have, but you are possibly skipping rows because of the `if at_least_three and less_than_four_condition:` in L97, so I added the slicing.